library/FrontRow/App/Http.hs

```diff
@@ -168,7 +173,38 @@ addAcceptHeader = addRequestHeader hAccept
 addBearerAuthorizationHeader :: BS.ByteString -> Request -> Request
 addBearerAuthorizationHeader = addRequestHeader hAuthorization . ("Bearer " <>)
 
--- | Read an 'Either' response body, throwing any 'Left' as an exception
+-- | Error thrown by 'getResponseBodyUnsafe' for unsuccessful status codes
+data HttpStatusError e a = HttpStatusError
+  { hseStatus :: Status
+  , hseResponse :: Response (Either e a)
+  }
+  deriving stock Show
+
+instance (Exception e, Typeable a, Show a) => Exception (HttpStatusError e a) where
+  displayException HttpStatusError {..} = unlines
+    [ "Received unsuccessful HTTP Status:"
+    , "Status: " <> show (statusCode hseStatus) <> " " <> BS8.unpack
+      (statusMessage hseStatus)
+    , case getResponseBody hseResponse of
+      Left e -> "Response body did not parse: " <> displayException e
+      Right a -> "Response body: " <> show a
```

Comment:
Just thinking out loud here, no changes expected, but if I use `httpJson @_ @Person`, and we get a non-200, it will still try to parse the body to a `Person`, but the body is almost certainly no longer a `Person`, right? We're going to get `Left`s unless the type in the body in the error case can be deserialized to the same type as in the success case. 

Completely minor point, just trying to make sure I am following this properly and not making stuff up.

Reply:
> but the body is almost certainly no longer a `Person`, right?

Yes that's exactly what would happen. My changes in this PR aren't really meant to change that, but just make the situation downstream of that a _little_ more convenient in _some_ cases.

I'm basing this on the premise that if that's the types you decided to use, that's the only reasonable behavior; and so, all we can really do is try to make it so you see a status-code error first, in cases where it would most likely bite you to not.

Since `getResponseBody` gives you control over the `Either`, you can/should/will be exercising the same control and looking at `getResponseStatus` beforehand, if desired; so I've left that as-is. Since `getResponseBodyUnsafe` is dispatching the `Either` for you, that's where I see the most risk, so that's where I'm injecting the automatic status-check.

But wait, what's the Right Way to address this situation? Well, you shouldn't parse to `Person`!

> unless the type in the body in the error case can be deserialized to the same type as in the success case

Hit the nail right on the head. Example:

```hs
-- However this API structures error messages
data Error = Error
  { code :: Int
  , message :: Text
  }
  deriving stock (Show, Generic)
  deriving anyclass FromJSON

data OrError a = Succeeded a | OrError Error

instance FromJSON a => FromJSON (OrError a) where
  parseJSON v = asum
    [ OrError <$> parseJSON v
    , Success <$> parseJSON v
    ]

main :: IO ()
main = do
  resp <- httpJson @_ @(OrError Person) someRequest

  case getResponseBody of
    Right (Success person) -> doSomethingWithPerson person

    Right (OrError e) | getResponseStatus resp == status400 ->
      logError $ "Bad request: " <> pack (show e)

    Right (OrError e) ->
      logError $ "Other error: " <> pack (show e)

    Left e ->
      logWarn $ "Unexpected response: " <> displayException e
```

I've been trying to come up with a way for `FrontRow.App.Http` to make this pattern easier, but it's so specific to the API in question that I can't.